docs/swift/DesignDoc.md

```diff
@@ -0,0 +1,674 @@
+easily support nesting submodules (Clang module maps support this, but pure
+Swift does not yet provide a way to define submodules).
+
+We will generate types with fully-qualified underscore-delimited names. For
+example, a message `Baz` in package `foo.bar` would generate a struct named
+`Foo_Bar_Baz`. For each fully-qualified proto message, there will be exactly one
+unique type symbol emitted in the generated binary.
+
+Users are likely to balk at the ugliness of underscore-delimited names for every
+generated type. To improve upon this situation, we will add a new string file
+level option, `swift_package_typealias`, that can be added to `.proto` files.
+When present, this will cause `typealias`es to be added to the generated Swift
+messages that replace the package name prefix with the provided string. For
+example, the following `.proto` file:
+
+```protobuf
+option swift_package_typealias = "FBP";
+package foo.bar;
+
+message Baz {
+  // Message fields
+}
+```
+
+would generate the following Swift source:
+
+```swift
+public struct Foo_Bar_Baz {
+  // Message fields and other methods
+}
+
+typealias FBPBaz = Foo_Bar_Baz
+```
+
+It should be noted that this type alias is recorded in the generated
+`.swiftmodule` so that code importing the module can refer to it, but it does
+not cause a new symbol to be generated in the compiled binary (i.e., we do not
+risk compiled size bloat by adding `typealias`es for every type).
+
+Other strategies to handle packages that were considered and rejected can be
+found in [Appendix A](#appendix-a-rejected-strategies-to-handle-packages).
+
+## Messages
+
+Proto messages are natural value types and we will generate messages as structs
+instead of classes. Users will benefit from Swift’s built-in behavior with
+regard to mutability. We will define a `ProtoMessage` protocol that defines the
+common methods and properties for all messages (such as serialization) and also
+lets users treat messages polymorphically. Any shared method implementations
+that do not differ between individual messages can be implemented in a protocol
+extension.
+
+The backing storage itself for fields of a message will be managed by a
+`ProtoFieldStorage` type that uses an internal dictionary keyed by field number,
+and whose values are the value of the field with that number (up-cast to Swift’s
+`Any` type). This class will provide type-safe getters and setters so that
+generated messages can manipulate this storage, and core serialization logic
+will live here as well. Furthermore, factoring the storage out into a separate
+type, rather than inlining the fields as stored properties in the message
+itself, lets us implement copy-on-write efficiently to support passing around
+large messages. (Furthermore, because the messages themselves are value types,
+inlining fields is not possible if the fields are submessages of the same type,
+or a type that eventually includes a submessage of the same type.)
```

Comment:
In fact, after implementing an early prototype, we've abandoned this approach (so this document is already somewhat out of date 😄 ). The idea here was to optimize for code size by making the storage and serialization logic as generic and reusable as possible, but the problem we ran into was significant performance penalties when working with repeated fields and maps. Specifically, Swift's copy-on-write behavior for arrays and dictionaries meant that whenever a collection needed to be retrieved to be mutated, the underlying storage was never uniquely referenced—the reference in the `ProtoFieldStorage` dictionary still existed. Thus, any mutation would create a copy, mutate the copy, and then store it back in PFS, throwing away the original.

Instead, we've moved toward a hybrid approach based on a previous design (not documented here) that performs much better. We leave arrays and dictionaries as properties defined directly in the generated struct, since they already implement their own CoW and making them directly accessible (i.e., being able to call `append` directly on a **stored** property) means no extraneous copies are being made. For singular fields, we have a private CoW storage class inside the message to store those, with computed properties on the message itself to access them. The idea here was to improve performance when working with messages that have a large number of singular fields by better controlling when copies of the storage are made.

We've considered investigating what the "sweet spot" is for when a message is small enough that we can eliminate the storage class and just use straight stored properties there as well, but haven't moved in that direction yet. I'll also note that `String` fields (and the Foundation `Data` value type in Swift 3) could be treated similarly to the repeated fields above in terms of CoW semantics, but that's still open as a refinement.)